scylla-cql/src/types/serialize/value.rs

```diff
@@ -283,30 +283,178 @@ impl<T: Value + ?Sized> SerializeCql for &T {
-    fallback_impl_contents!();
+impl<T: SerializeCql> SerializeCql for Vec<T> {
+    fn preliminary_type_check(typ: &ColumnType) -> Result<(), SerializationError> {
+        match typ {
+            ColumnType::List(elt) | ColumnType::Set(elt) => {
+                T::preliminary_type_check(elt).map_err(|err| {
+                    mk_typck_err::<Self>(
+                        typ,
+                        SetOrListTypeCheckErrorKind::ElementTypeCheckFailed(err),
+                    )
+                })
+            }
+            _ => Err(mk_typck_err::<Self>(
+                typ,
+                SetOrListTypeCheckErrorKind::NotSetOrList,
+            )),
+        }
+    }
+
+    fn serialize<W: CellWriter>(
+        &self,
+        typ: &ColumnType,
+        writer: W,
+    ) -> Result<W::WrittenCellProof, SerializationError> {
+        serialize_sequence(
+            std::any::type_name::<Self>(),
+            self.len(),
+            self.iter(),
+            typ,
+            writer,
+        )
+    }
 }
-impl<T: Value> SerializeCql for &[T] {
-    fallback_impl_contents!();
+impl<'a, T: SerializeCql + 'a> SerializeCql for &'a [T] {
+    fn preliminary_type_check(typ: &ColumnType) -> Result<(), SerializationError> {
+        match typ {
+            ColumnType::List(elt) | ColumnType::Set(elt) => {
+                T::preliminary_type_check(elt).map_err(|err| {
+                    mk_typck_err::<Self>(
+                        typ,
+                        SetOrListTypeCheckErrorKind::ElementTypeCheckFailed(err),
+                    )
+                })
+            }
+            _ => Err(mk_typck_err::<Self>(
+                typ,
+                SetOrListTypeCheckErrorKind::NotSetOrList,
+            )),
+        }
+    }
+
+    fn serialize<W: CellWriter>(
+        &self,
+        typ: &ColumnType,
+        writer: W,
+    ) -> Result<W::WrittenCellProof, SerializationError> {
+        serialize_sequence(
+            std::any::type_name::<Self>(),
+            self.len(),
+            self.iter(),
+            typ,
+            writer,
+        )
+    }
 }
 impl SerializeCql for CqlValue {
     fallback_impl_contents!();
 }
 
 fallback_tuples!(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15);
 
+fn serialize_sequence<'t, T: SerializeCql + 't, W: CellWriter>(
+    rust_name: &'static str,
+    len: usize,
+    iter: impl Iterator<Item = &'t T>,
+    typ: &ColumnType,
+    writer: W,
+) -> Result<W::WrittenCellProof, SerializationError> {
+    let elt = match typ {
+        ColumnType::List(elt) | ColumnType::Set(elt) => elt,
+        _ => {
+            return Err(mk_typck_err_named(
+                rust_name,
+                typ,
+                SetOrListTypeCheckErrorKind::NotSetOrList,
+            ));
+        }
+    };
+
+    let mut builder = writer.into_value_builder();
+
+    let element_count: i32 = len.try_into().expect("too many values to fit in i32");
```

Comment:
Hmm...

The interface I introduced in https://github.com/scylladb/scylla-rust-driver/pull/855 panics when we try to produce a value which overflows `i32`. Panicking here would be consistent with that (each item has at least 4 bytes, so the whole thing overflows). I know that the old interface does not do this, but my rationale back then was that

- `Vec::push` panics on overflow,
- In one of the iterations of the design, I was updating serialized length in destructors, and that could overflow - the only sensible way to handle that was to panic.

However:

- `Vec` uses `isize` and it's not possible currently to overflow it without running out of memory; it's not true for `i32`,
- The current design doesn't have custom destructor logic and `finish()` could be used to report the overflow error.

What do you think? Personally, I'm changing my mind about panicking and I think I should go over and remove the panic-on-overflow logic.

Reply:
I think that the 2 occurences of `expect` in this files can be removed (because they happen in functions returning `SerializationError`), so they should be removed - because as you said previous interface didn't panic, but also just to avoid panics in a library code.

Reply:
I'll remove the two `expect`s, but I'll de-panic the row writers / cell writers / cell value builders. Some of their methods will start returning `Result` so there will be a large number of changes to this PR, but they will mostly be mechanical.

Reply:
This is now fixed in the other PR.